onnx/test/cpp/data_propagation_test.cc

```diff
@@ -0,0 +1,156 @@
+                A->dim(i).dim_value() != B->dim(i).dim_value()) {
+            fail_check("The compared dim values are different.");
+            return false;
+        }
+    }
+    return true;
+}
+
+void RunDataPropagationTest(const char* graphCode, int domain_version = 15) {
+  // Parses the graph from graphCode
+  GraphProto graph;
+  OnnxParser parser(graphCode);
+  auto status = parser.Parse(graph);
+  EXPECT_TRUE(status.IsOK()) << status.ErrorMessage();
+  EXPECT_TRUE(parser.EndOfInput()) << "Extra unparsed input unexpected.";
+
+  // Constructs name to TypeProto map from value_info, input, output
+  std::unordered_map<std::string, TypeProto*> valueTypesByName;
+  for (auto& vi : *graph.mutable_value_info()) {
+    if (vi.has_type()) {
+      valueTypesByName[vi.name()] = vi.mutable_type();
+    }
+  }
+  for (auto& vi : *graph.mutable_input()) {
+    if (vi.has_type()) {
+      valueTypesByName[vi.name()] = vi.mutable_type();
+    }
+  }
+  for (auto& vi : *graph.mutable_output()) {
+    if (vi.has_type()) {
+      valueTypesByName[vi.name()] = vi.mutable_type();
+    }
+  }
+
+  // Constructs name to TensorProto map from initializer
+  std::unordered_map<std::string, const TensorProto*> inputDataByName;
+  for (const auto& tp : graph.initializer()) {
+    inputDataByName[tp.name()] = &tp;
+  }
+  // Collects data from constant nodes
+  for (const auto& n : graph.node()) {
+    if (n.op_type() != "Constant" || n.output().size() != 1) {
+      continue;
+    }
+    for (const auto& attr : n.attribute()) {
+      if (attr.name() == "value") {
+        if (attr.type() == AttributeProto::TENSOR && attr.has_t()) {
+          inputDataByName[n.output(0)] = &attr.t();
+        }
+      }
+    }
+  }
+
+  // Runs data propagation on each node
+  std::unordered_map<std::string, TensorShapeProto> generatedShapeDataByName;
+  auto* schemaRegistry = OpSchemaRegistry::Instance();
+  const TensorShapeProto* propagatedShape;
+  for (auto n: graph.node()) {
+    // No need to run data propagation on Constant
+    if (n.op_type() == "Constant") {
+      continue;
+    }
+    DataPropagationContextImpl dataPropagationCtx(
+        n, valueTypesByName, inputDataByName, generatedShapeDataByName);
+    const auto schema = schemaRegistry->GetSchema(n.op_type(), domain_version, n.domain());
+    EXPECT_TRUE(schema->has_data_propagation_function());
+    schema->GetDataPropagationFunction()(dataPropagationCtx);
+    propagatedShape = dataPropagationCtx.getInputData(0);
+  }
+  // Expects the input data of final node (from the output of previous node)
+  // is same as the given output shape
+  auto* outputShape = graph.mutable_output(0)
+    ->mutable_type()->mutable_tensor_type()->mutable_shape();
+  EXPECT_TRUE(CompareShape(propagatedShape, outputShape));
+}
+
+TEST(DataPropagationImplTest, ShapeTest) {
+  const char* code = R"ONNX(
+agraph (int32[7,4,1] x) => (int32[7,4,1] y)
+{
+    xs = Shape(x)
+    y = Cast<to = 7>(xs)
+}
```

Comment:
this graph is invalid...
Shape of x is [7, 4, 1] this is correct however the shape of xs and y will not be [7, 4, 1] it will be [3] because shaoe node will output [7, 4,3] and its shape will be [3]... similarly cast will simply cast the data without changing the shape so shape of y will be [3]

Your test passes because you are not calling TypeAndShapeInferenceFunction for the node before calling DataPropagationFunction

Also check on line 105 is not right... that function should simply return the computed shape data... every individual node test should run comparision... for nodes like shape you will need to hard code the expected output